internal/tiltfile/flags/args_def.go

```diff
@@ -2,66 +2,155 @@ package flags
 
 import (
 	"bytes"
-	"errors"
+	"encoding/json"
 	"flag"
 	"fmt"
+	"os"
 
+	jsoniter "github.com/json-iterator/go"
+	"github.com/pkg/errors"
 	"go.starlark.net/starlark"
 
 	"github.com/windmilleng/tilt/internal/tiltfile/starkit"
+	"github.com/windmilleng/tilt/pkg/model"
 )
 
 type argValue interface {
-	flag() flag.Value
+	flag.Value
+	json.Marshaler
 	starlark() starlark.Value
 	setFromArgs([]string)
+	setFromInterface(interface{}) error
+	IsSet() bool
 }
 
 type argDef struct {
-	argValue
-	usage string
+	newValue func() argValue
+	usage    string
 }
 
 type ArgsDef struct {
 	positionalArgName string
 	args              map[string]argDef
 }
 
-func (ad ArgsDef) parse(args []string) (v starlark.Value, output string, err error) {
+func (ad ArgsDef) parse(flagsState model.FlagsState, args []string) (v starlark.Value, mergedArgs bool, output string, err error) {
+	var config map[string]argValue
+	config, err = ad.readFromFile(flagsState.ConfigPath)
+	if err != nil {
+		return starlark.None, false, "", err
+	}
+
+	// if we have not yet merged the current set of args, merge them into the flags from the file
+	// and write them back out
+	if flagsState.LastArgsWrite.IsZero() {
+		var flagsFromArgs map[string]argValue
+		flagsFromArgs, output, err = ad.parseArgs(args)
+		if err != nil {
+			return nil, false, output, err
+		}
+
+		for k, v := range flagsFromArgs {
+			if v.IsSet() {
+				config[k] = v
+			}
+		}
+
+		f, err := os.Create(flagsState.ConfigPath)
+		if err != nil {
+			return nil, false, output, errors.Wrapf(err, "error opening %s for writing", flagsState.ConfigPath)
+		}
+		defer func() {
+			err2 := f.Close()
+			if err2 != nil && err == nil {
+				err = errors.Wrapf(err2, "error closing %s", flagsState.ConfigPath)
```

Comment:
i don't think this does what you expect?

this is setting the `err` variable declared inside the `if` block, not the variable declared as a return value. here's an example:
https://play.golang.org/p/0fTWw4vsdPA

in general, the "set a named return value from inside a defer" is pretty error-prone and has lots of little gotchas like this. I'm not sure if there's much value in handling Close() errors at all

Reply:
whoops, good catch